### PR TITLE
Implement tests

### PR DESCRIPTION
**Pull request by ilisau**
In this PR I implemented tests

---

## PR-Codex overview
This PR adds some new dependencies and introduces new classes for fake services, as well as some changes to existing code, such as adding a new parameter to a method and removing a line of code. 

### Detailed summary
- Added `reactor-test` dependency to `pom.xml`
- Added new classes for fake services: `FakeAuthService`, `FakeTourService`, `FakeJwtService`, and `FakeUserClient`
- Added package-info.java file to `src/main/java/com/solvd/qaprotours/service/impl/fake/`
- Added a new parameter to `AuthServiceImpl.restoreUserPassword()` method: `params.put("user.id", user.getId());`
- Removed `.then()` from `TicketServiceImpl.flatMap(tour -> ticketRepository.deleteById(ticketId))` method call
- Added `SchedulerTests` and `ImageServiceTests` to `src/test/java/com/solvd/qaprotours/service/impl/`

> The following files were skipped due to too many changes: `src/test/java/com/solvd/qaprotours/service/impl/ImageServiceTests.java`, `src/main/java/com/solvd/qaprotours/service/impl/fake/FakeTicketService.java`, `src/test/java/com/solvd/qaprotours/config/TestConfig.java`, `src/test/java/com/solvd/qaprotours/service/impl/TourServiceTests.java`, `src/test/java/com/solvd/qaprotours/service/impl/TicketServiceTests.java`, `src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        StepVerifier.create(result)
+                .expectNext(response)
+                .verifyComplete();
+        verify(jwtService).generateToken(JwtTokenType.ACCESS, user);
+        verify(jwtService).generateToken(JwtTokenType.REFRESH, user);
+        verify(passwordEncoder).matches(eq(authentication.getPassword()), anyString());
+    }
+
+    @Test
+    void loginWithIncorrectPassword() {
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java:106]: @ilisau avoid those long tests. [read](https://www.yegor256.com/2017/05/17/single-statement-unit-tests.html) about single-statement test technique

File: src/test/java/com/solvd/qaprotours/service/impl/ImageServiceTests.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            Mono<Void> result = imageService.uploadImage(1L, image);
+            StepVerifier.create(result)
+                    .expectNextCount(0)
+                    .verifyComplete();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static InputStream generateImage() throws Exception {
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/service/impl/ImageServiceTests.java:82]: @ilisau what is it?

File: src/test/java/com/solvd/qaprotours/service/impl/SchedulerTests.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        when(mailDataMapper.toDto(any()))
+                .thenReturn(new MailDataDto());
+        when(messageSender.sendMessage(eq("mail"), anyInt(), anyString(), any()))
+                .thenReturn(Flux.empty());
+        scheduler.findBookedTickets();
+        verify(messageSender, times(tickets.size())).sendMessage(eq("mail"), anyInt(), anyString(), any());
+    }
+
+    @Test
+    void findNotConfirmedTickets() {
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/service/impl/SchedulerTests.java:97]: @ilisau absolutely unmaintainable tests, fix it.
also, consider [reading](https://www.yegor256.com/2018/12/11/unit-testing-anti-patterns.html) about test patterns and anti-patterns

**Review by h1alexbel — COMMENTED** [review]
@ilisau pull request is too large, hard to review it

**Comment by ilisau** [comment]
@h1alexbel please review this PR

File: src/test/java/com/solvd/qaprotours/service/impl/SchedulerTests.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        )
+                .thenReturn(Flux.empty());
+        scheduler.findBookedTickets();
+        verify(messageSender, times(tickets.size()))
+                .sendMessage(eq("mail"), anyInt(), anyString(), any());
+    }
+
+    @Test
+    void findNotConfirmedTickets() {
+        User user = generateUser();
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/service/impl/SchedulerTests.java:78]: the same

**Review by h1alexbel — CHANGES_REQUESTED** [review]

File: src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.test.StepVerifier;
+
+import static org.mockito.ArgumentMatchers.any;
```
[the remark this answers is not in the document]
Comment (ilisau) [reply src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java:27]: default imports are not available there

File: src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java [hunk 6]
```diff
@@ -66,7 +62,9 @@ void loginWithCorrectPassword() {
                 .thenReturn(Mono.just(userDto));
         when(userMapper.toEntity(userDto))
                 .thenReturn(user);
-        when(passwordEncoder.matches(eq(authentication.getPassword()), anyString()))
+        when(passwordEncoder.matches(
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java:65]: @ilisau static imports should be removed

**Review by h1alexbel — CHANGES_REQUESTED** [review]

File: src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+    @Mock
+    private MessageSender messageSender;
+
+    @InjectMocks
+    private AuthServiceImpl authService;
+
+    @Test
+    void loginWithCorrectPassword() {
+        User user = generateUser();
```
[the remark this answers is not in the document]
Comment (h1alexbel) [reply src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java:57]: @ilisau you don't need to mock inside the service logic using Mockito or any other frameworks

you just need to obey your interface contract and return something fake
this is the main idea behind fake objects

File: src/main/java/com/solvd/qaprotours/web/security/SecurityConfig.java [hunk 8]
```diff
@@ -63,7 +61,7 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
                 .requestMatchers("/swagger-ui/**").permitAll()
                 .requestMatchers("/v3/api-docs/**").permitAll()
                 .requestMatchers("/actuator/**").permitAll()
-                .anyRequest().authenticated()
+                .anyRequest().permitAll()
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/web/security/SecurityConfig.java:64]: @ilisau I think the actuator endpoints need to be hidden from anonymous accounts, and only users with role `ADMIN` have ability to use those endpoints

WDYT?

File: src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java [hunk 9]
```diff
@@ -58,17 +56,17 @@ void loginWithCorrectPassword() {
         UserDto userDto = generateUserDto();
         Authentication authentication = generateAuthentication(user);
         JwtResponse response = generateResponse();
-        when(userClient.getByEmail(authentication.getEmail()))
+        Mockito.when(userClient.getByEmail(authentication.getEmail()))
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java:59]: @ilisau why do you need to mock your already "faked" dependencies?

Using fake objects all you need is to inject fakes into object under test

File: src/main/java/com/solvd/qaprotours/web/security/SecurityConfig.java (hunk 8 above)
Comment (ilisau) [reply src/main/java/com/solvd/qaprotours/web/security/SecurityConfig.java:64]: @h1alexbel circuitbreaker calls this api to check whether service is alive, so it is complicated to generate tokens for this feature

Comment (h1alexbel) [reply src/main/java/com/solvd/qaprotours/web/security/SecurityConfig.java]: @ilisau yes, you are right, but how to make this endpoint hidden from other users?

**Comment by h1alexbel** [comment]
> Spring does not inject dependencies anyway in tests. And in some tests I need to mock repositories to interact with them. Otherwise, if I use `@Mock` and fake service without `Mockito.when(...)` it returns null in all methods, not what it actually should return in implemented fake services

Mocking the repositories is OK

File: src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java [hunk 10]
```diff
@@ -36,22 +38,16 @@ public class AuthServiceTests {
     @Mock
     private UserClient userClient;
 
-    @Mock
+    @MockBean
     private BCryptPasswordEncoder passwordEncoder;
 
-    @Mock
+    @MockBean
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java:44]: @ilisau nice

File: src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java [hunk 11]
```diff
@@ -121,19 +130,26 @@ void sendRestoreToken() {
...
     void restoreUserPasswordWithCorrectToken() {
         String userId = "1";
-        String token = "token";
+        String token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWQiOiIxIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.19oTkCTJL9PZRNk3Z8bao3UdQNTX2d7fdw2ijDaW-iI";
         String password = "password";
+        Claims claims = new DefaultClaims();
+        claims.put("id", "1");
+        Mockito.when(jwtService.validateToken(ArgumentMatchers.anyString()))
+                .thenReturn(true);
+        Mockito.when(jwtService.isTokenType(ArgumentMatchers.anyString(), ArgumentMatchers.eq(JwtTokenType.RESET)))
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/service/impl/AuthServiceTests.java:145]: @ilisau So, here you're mocking only jwt service

**Review by h1alexbel — APPROVED** [review]
@ilisau a little more maintainable)

**Review by h1alexbel — APPROVED** [review]

**Review by h1alexbel — APPROVED** [review]

**Comment by ilisau** [comment]
@h1alexbel So now I need to create repositories` beans, because I also need to create tested services beans

File: src/test/java/com/solvd/qaprotours/config/TestConfig.java [hunk 12]
```diff
@@ -1,32 +1,104 @@
...
 
-    private final JwtProperties jwtProperties;
+    private final TicketRepository ticketRepository;
+    private final TourRepository tourRepository;
+
+    @Bean
+    public JwtProperties jwtProperties() {
+        return new JwtProperties(
+                1,
+                24,
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/config/TestConfig.java:50]: @ilisau what is the 24?

File: src/test/java/com/solvd/qaprotours/config/TestConfig.java [hunk 13]
```diff
@@ -1,32 +1,104 @@
...
+    @Primary
+    @Bean
+    public Scheduler scheduler() {
+        return new Scheduler(ticketService(), userClient(), mailDataMapper(), userMapper(), messageSender());
+    }
+
+    @Primary
+    @Bean
+    public MessageSender messageSender() {
+        return null;
```
Comment (h1alexbel) [inline src/test/java/com/solvd/qaprotours/config/TestConfig.java:100]: @ilisau don't return NULL

avoid this technique at all costs

File: src/test/java/com/solvd/qaprotours/config/TestConfig.java (hunk 12 above)
Comment (ilisau) [reply src/test/java/com/solvd/qaprotours/config/TestConfig.java:50]: time constants from application.yml

**Pull request merged**